### PR TITLE
FINERACT-2081: Apply penalty to overdue loans enhancements

### PR DESCRIPTION
**Pull request by alberto-art3ch**
## Description

Describe the changes made and why they were made.

- When there is no overdue penalty available and no overdue installment, do not reprocess the loan  

- Enhance `retrieveAllOverdueInstallmentsForLoan`  logic, if there is no penalty charge configured on the loan, do not iterate through on the installments

[FINERACT-2081](https://issues.apache.org/jira/browse/FINERACT-2081)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/ApplyChargeToOverdueLoansBusinessStep.java [hunk 1]
```diff
@@ -37,7 +37,9 @@ public class ApplyChargeToOverdueLoansBusinessStep implements LoanCOBBusinessSte
     public Loan execute(Loan loan) {
         final Collection<OverdueLoanScheduleData> overdueLoanScheduleDataList = loanReadPlatformService
                 .retrieveAllOverdueInstallmentsForLoan(loan);
-        loanChargeWritePlatformService.applyOverdueChargesForLoan(loan.getId(), overdueLoanScheduleDataList);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/loan/ApplyChargeToOverdueLoansBusinessStep.java:40]: Please dont forget to enhance the "Apply Penalty to overdue loans" job as well!

Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/cob/loan/ApplyChargeToOverdueLoansBusinessStep.java:40]: Done! job code updated

File: fineract-e2e-tests-runner/src/test/resources/features/LoanDelinquency.feature [hunk 2]
```diff
@@ -625,10 +625,7 @@ Feature: LoanDelinquency
       | PAUSE  | 15 October 2023 | 30 October 2023 |
     Then Loan has the following LOAN level delinquency data:
       | classification | delinquentAmount | delinquentDate  | delinquentDays | pastDueDays |
-      | RANGE_3        | 250.0            | 04 October 2023 | 11             | 14          |
-    Then Loan has the following INSTALLMENT level delinquency data:
-      | rangeId | Range   | Amount |
-      | 2       | RANGE_3 | 250.00 |
+      | NO_DELINQUENCY | 250.0            | 04 October 2023 | 11             | 14          |
```
Comment (adamsaghy) [inline fineract-e2e-tests-runner/src/test/resources/features/LoanDelinquency.feature:628]: I think these tests should be fixed to run the delinquency business step properly and not by changing the outcome. Would you consider discuss these things with @peter-kovacs-dpc ?

**Review by galovics — CHANGES_REQUESTED** [review]
Tests are failing. Please fix them. Thanks.

**Comment by galovics** [comment]
@alberto-art3ch do you plan on fixing this?

**Comment by adamsaghy** [comment]
Already been fixed by other PR.

**Pull request closed without merging**